miniwin/src/ddraw/ddraw.cpp

```diff
@@ -323,25 +294,44 @@ HRESULT DirectDraw_SDL3GPUImpl::SetCooperativeLevel(HWND hWnd, DDSCLFlags dwFlag
 		if (!SDL_SetWindowFullscreen(hWnd, fullscreen)) {
 			return DDERR_GENERIC;
 		}
-		DDWindow_SDL3GPU = hWnd;
+		DDWindow = hWnd;
 	}
 	return DD_OK;
 }
 
-HRESULT DirectDraw_SDL3GPUImpl::SetDisplayMode(DWORD dwWidth, DWORD dwHeight, DWORD dwBPP)
+HRESULT DirectDrawImpl::SetDisplayMode(DWORD dwWidth, DWORD dwHeight, DWORD dwBPP)
 {
 	MINIWIN_NOT_IMPLEMENTED();
 	return DD_OK;
 }
 
 // IDirect3D2 interface
-HRESULT DirectDraw_SDL3GPUImpl::CreateDevice(const GUID& guid, void* pBackBuffer, IDirect3DDevice2** ppDirect3DDevice)
+HRESULT DirectDrawImpl::CreateDevice(
+	const GUID& guid,
+	IDirectDrawSurface* pBackBuffer,
+	IDirect3DDevice2** ppDirect3DDevice
+)
 {
-	*ppDirect3DDevice = new IDirect3DDevice2;
+	DDSURFACEDESC DDSDesc;
+	DDSDesc.dwSize = sizeof(DDSURFACEDESC);
+	pBackBuffer->GetSurfaceDesc(&DDSDesc);
+
+	Direct3DRMRenderer* renderer;
+	if (SDL_memcmp(&guid, &SDL3_GPU_GUID, sizeof(GUID)) == 0) {
+		renderer = Direct3DRMSDL3GPURenderer::Create(DDSDesc.dwWidth, DDSDesc.dwHeight);
+	}
+	else if (SDL_memcmp(&guid, &SOFTWARE_GUID, sizeof(GUID)) == 0) {
+		renderer = new Direct3DRMSoftwareRenderer(DDSDesc.dwWidth, DDSDesc.dwHeight);
+	}
+	else {
+		SDL_LogError(SDL_LOG_CATEGORY_APPLICATION, "Device GUID not recognized");
+		return E_NOINTERFACE;
+	}
+	*ppDirect3DDevice = static_cast<IDirect3DDevice2*>(renderer);
 	return DD_OK;
 }
 
-HRESULT DirectDrawEnumerate_SDL3GPU(LPDDENUMCALLBACKA cb, void* context)
+HRESULT DirectDrawEnumerate(LPDDENUMCALLBACKA cb, void* context)
```

Comment:
Current start-up behavior of sdl3gpu is imho horrible: the sdl3gpu driver is created and destroyed once for each video driver.
`DirectDrawEnumerate` does not make sense currently: `SDL_Init(SDL_INIT_VIDEO)` has already been called at this point, so the video driver is already fixed.

So, this function should execute the callback only once, and use [`SDL_GetCurrentVideoDriver`](https://wiki.libsdl.org/SDL3/SDL_GetCurrentVideoDriver).